### PR TITLE
Add `BoundedBTreeMap` to `frame_support::storage`

### PR DESCRIPTION
**Pull request by coriolinus**
Part of https://github.com/paritytech/substrate/issues/8719.

Tests ~~forthcoming~~ implemented.

### CODE REVIEW

**Comment by shawntabrizi** [comment]
Initial review looks good 👍

File: frame/support/src/storage/bounded_btree_map.rs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+	pub fn remove_entry<Q>(&mut self, key: &Q) -> Option<(K, V)>
+	where
+		K: Borrow<Q>,
+		Q: Ord + ?Sized,
+	{
+		self.0.remove_entry(key)
+	}
+}
+
+impl<K, V, S> Default for BoundedBTreeMap<K, V, S>
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_btree_map.rs:171]: Can't you use `DefaultNoBound` and similar?

File: frame/support/src/storage/bounded_btree_map.rs [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+	/// # use frame_support::{parameter_types, storage::bounded_btree_map::BoundedBTreeMap};
+	/// parameter_types! {
+	/// 	pub const Size: u32 = 5;
+	/// }
+	/// let mut map = BTreeMap::new();
+	/// map.insert("foo", 1);
+	/// map.insert("bar", 2);
+	/// let bounded_map = unsafe {BoundedBTreeMap::<_, _, Size>::force_from(map, "demo")};
+	/// ```
+	pub unsafe fn force_from<Scope>(map: BTreeMap<K, V>, scope: Scope) -> Self
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_btree_map.rs:82]: If this is unsafe, the one from `BoundedVec` should also be unsafe.

File: frame/support/src/storage/bounded_btree_map.rs [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+impl<K, V, S> Clone for BoundedBTreeMap<K, V, S>
+where
+	BTreeMap<K, V>: Clone,
+{
+	fn clone(&self) -> Self {
+		BoundedBTreeMap(self.0.clone(), PhantomData)
+	}
+}
+
+impl<K, V, S> fmt::Debug for BoundedBTreeMap<K, V, S>
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_btree_map.rs:190]: You might want to strip some of this if `cfg[feature = "std"]`

File: frame/support/src/storage/bounded_btree_map.rs (hunk 1 above)
Comment (coriolinus) [reply frame/support/src/storage/bounded_btree_map.rs:171]: No, unfortunately; I didn't look into the root cause in much detail, but when I attempted it, it seemed that that macro doesn't produce appropriate bounds for this case; it just generated a bunch of compile errors. The manual implementation was simple enough anyway.

**Review by kianenigma — APPROVED** [review]

File: frame/support/src/storage/bounded_btree_map.rs (hunk 2 above)
Comment (coriolinus) [reply frame/support/src/storage/bounded_btree_map.rs:82]: 21b025ec1

File: frame/support/src/storage/bounded_btree_map.rs (hunk 3 above)
Comment (coriolinus) [reply frame/support/src/storage/bounded_btree_map.rs:191]: e6662021f

**Review by shawntabrizi — APPROVED** [review]

**Comment by coriolinus** [comment]
bot merge

**Pull request merged**